rust/actyx/ax-aql/src/language/type_check/drill_down_type.rs

```diff
@@ -0,0 +1,443 @@
+                    return drill_down(&collapsed, index);
+                }
+            };
+
+            // index pop_front-ing
+            index = rest;
+        }
+    }
+
+    fn drill_down_with_calculated_expressions(
+        current_type: &Type,
+        first_expr: &SimpleExpr,
+        rest: &[Index],
+    ) -> Result<Type, String> {
+        let new_index = match first_expr {
+            SimpleExpr::Number(Num::Natural(x)) => Index::Number(*x),
+            SimpleExpr::String(x) => Index::String(x.clone()),
+            SimpleExpr::Number(Num::Decimal(_)) => {
+                return Err(format!(
+                    "{:?} cannot be accessed by {}. Decimal cannot be used for indexing.",
+                    current_type, first_expr
+                ))
+            }
+            SimpleExpr::Bool(_)
+            | SimpleExpr::Object(_)
+            | SimpleExpr::Array(_)
+            | SimpleExpr::Null
+            | SimpleExpr::Not(_)
+            | SimpleExpr::KeyVar(_)
+            | SimpleExpr::KeyLiteral(_)
+            | SimpleExpr::TimeVar(_)
+            | SimpleExpr::TimeLiteral(_)
+            | SimpleExpr::Tags(_)
+            | SimpleExpr::App(_)
+            | SimpleExpr::BinOp(_) => return Err(format!("{:?} cannot be accessed by {}", current_type, first_expr)),
+            SimpleExpr::Indexing(_) => {
+                // TODO: support
+                return Err(format!(
+                    "{:?} cannot be accessed by {:?}. Indexing by an indexing expression isn't supported yet",
+                    current_type, first_expr
+                ));
+            }
+            SimpleExpr::Variable(_)
+            | SimpleExpr::Interpolation(_)
+            | SimpleExpr::AggrOp(_)
+            | SimpleExpr::FuncCall(_)
+            | SimpleExpr::SubQuery(_) => return drill_down(&Type::Atom(TypeAtom::Universal), rest),
+            SimpleExpr::Cases(c) => {
+                // Case acts similarly to a union because it may results in several types
+
+                let type_results = c
+                    .iter()
+                    .map(|(_, then)| drill_down_with_calculated_expressions(current_type, then, rest))
+                    .collect::<Vec<_>>();
+
+                let errors = type_results
+                    .iter()
+                    .filter_map(|x| match x {
+                        Ok(_) => None,
+                        Err(err_msg) => Some(err_msg),
+                    })
+                    .collect::<Vec<_>>();
+
+                if !errors.is_empty() {
+                    let error_messages = errors
+                        .into_iter()
+                        .map(|err_msg| format!("- {}", err_msg))
+                        .collect::<Vec<_>>()
+                        .join("\n");
+                    return Err(error_messages.trim().into());
+                }
+
+                // NOTE: Using Option to emulate bottom type; consider introducing bottom type
```

Comment:
Is `NO_VALUE` the bottom type? Maybe this could be addressed with it?

Reply:
no, no_value is not the subtype of NULL or NUMBER for example